// xadmin-antd@3.3.2/lib/layout/App.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _xadmin = require("xadmin");

var _antd = require("antd");

var _icons = require("@ant-design/icons");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var {
  Header,
  Content,
  Footer,
  Sider
} = _antd.Layout;
var SubMenu = _antd.Menu.SubMenu;

function itemRender(route, params, routes, paths) {
  var last = routes.indexOf(route) === routes.length - 1;

  var click = to => () => _xadmin.app.go(to);

  return last ? /*#__PURE__*/_react.default.createElement("span", null, route.breadcrumbName) : /*#__PURE__*/_react.default.createElement("a", {
    onClick: click('/' + paths.join(''))
  }, route.breadcrumbName);
}

class MenuBar extends _react.default.Component {
  render() {
    return /*#__PURE__*/_react.default.createElement(_xadmin.Block, {
      name: "main.menu"
    }, items => /*#__PURE__*/_react.default.createElement(_antd.Menu, {
      theme: "dark",
      mode: this.props.mode
    }, items));
  }

}

class App extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      collapsed: false,
      mode: 'inline'
    });

    _defineProperty(this, "toggle", () => {
      this.setState({
        collapsed: !this.state.collapsed,
        mode: !this.state.collapsed ? 'vertical' : 'inline'
      });
    });
  }

  render() {
    var {
      routes,
      params
    } = this.props;
    var Icon = this.state.collapsed ? _icons.MenuFoldOutlined : _icons.MenuUnfoldOutlined;
    return /*#__PURE__*/_react.default.createElement(_antd.Layout, {
      style: {
        minHeight: '100vh'
      }
    }, /*#__PURE__*/_react.default.createElement(Sider, {
      trigger: null,
      collapsible: true,
      collapsed: this.state.collapsed,
      style: {
        overflow: 'auto',
        height: '100vh',
        position: 'fixed',
        left: 0
      }
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "logo"
    }, (0, _xadmin.config)('site.logo'), (0, _xadmin.config)('site.title', 'Admin')), /*#__PURE__*/_react.default.createElement(MenuBar, {
      mode: this.state.mode
    })), /*#__PURE__*/_react.default.createElement(_antd.Layout, {
      style: {
        marginLeft: this.state.collapsed ? 80 : 200,
        transition: 'margin .2s'
      }
    }, /*#__PURE__*/_react.default.createElement(Header, {
      style: {
        background: '#fff',
        boxShadow: 'rgba(0, 0, 0, 0.25) 0px 1px 3px',
        zIndex: 10
      }
    }, /*#__PURE__*/_react.default.createElement(Icon, {
      className: "trigger pull-left",
      onClick: this.toggle,
      style: {
        lineHeight: 'inherit',
        marginLeft: -40,
        fontSize: '1.2rem'
      }
    }), /*#__PURE__*/_react.default.createElement("div", {
      style: {
        float: 'right'
      }
    }, /*#__PURE__*/_react.default.createElement(_xadmin.Block, {
      name: "top.right"
    }, items => /*#__PURE__*/_react.default.createElement(_antd.Menu, {
      mode: "horizontal",
      style: {
        lineHeight: '62px'
      }
    }, items)))), /*#__PURE__*/_react.default.createElement(Content, null, this.props.children), /*#__PURE__*/_react.default.createElement(Footer, {
      style: {
        textAlign: 'center'
      }
    }, "\xA9 ", /*#__PURE__*/_react.default.createElement("slot", null, (0, _xadmin.config)('site.copyright')))));
  }

}

var _default = App;
exports.default = _default;
