// xadmin-antd@3.3.2/lib/auth/components/CaptchaCodeInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var _xadmin = require("xadmin");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

class CaptchaCodeInput extends _react.default.Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      url: this.getCodeUrl()
    };
  }

  getCodeUrl() {
    var {
      field
    } = this.props;
    return (0, _xadmin.api)({}).host + (field.captcha_url || '/get_captcha_code') + '?random=' + Math.random().toString();
  }

  render() {
    var {
      input,
      field
    } = this.props;
    var {
      _t
    } = _xadmin.app.context;
    return /*#__PURE__*/_react.default.createElement(_antd.Row, {
      gutter: 8
    }, /*#__PURE__*/_react.default.createElement(_antd.Col, {
      span: 14
    }, /*#__PURE__*/_react.default.createElement(_antd.Input, _extends({}, input, field.attrs))), /*#__PURE__*/_react.default.createElement(_antd.Col, {
      span: 10
    }, /*#__PURE__*/_react.default.createElement(_antd.Tooltip, {
      title: _t('Click to refresh captcha code')
    }, /*#__PURE__*/_react.default.createElement("img", {
      style: {
        cursor: 'pointer',
        width: '100%'
      },
      onClick: () => this.setState({
        url: this.getCodeUrl()
      }),
      src: this.state.url
    }))));
  }

}

exports.default = CaptchaCodeInput;
