// xadmin-antd@3.3.2/lib/auth/components/Signin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _xadmin = _interopRequireWildcard(require("xadmin"));

var _antd = require("antd");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SigininLayout = (_ref) => {
  var {
    error,
    children,
    invalid,
    handleSubmit,
    submitting
  } = _ref;
  var {
    _t
  } = _xadmin.default.context;

  var {
    auth
  } = _xadmin.default.get('config');

  var nav = (0, _xadmin.use)('navigate');
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_antd.Form, {
    className: "sigin-form",
    onSubmit: handleSubmit
  }, /*#__PURE__*/_react.default.createElement(_antd.Card, {
    style: {
      maxWidth: 450,
      margin: '5rem auto',
      marginBottom: '-0.5rem'
    },
    title: _t('Please Login')
  }, children, error && /*#__PURE__*/_react.default.createElement(_antd.Alert, {
    message: error,
    type: "error"
  }), /*#__PURE__*/_react.default.createElement(_antd.Form.Item, {
    wrapperCol: {
      xs: {
        span: 24
      },
      sm: {
        span: 18,
        offset: 5
      }
    }
  }, auth.can_reset_password && /*#__PURE__*/_react.default.createElement("div", null, _t('Forgot password'), "? ", /*#__PURE__*/_react.default.createElement("a", {
    onClick: () => nav('/forget_password')
  }, _t('reset password'))), /*#__PURE__*/_react.default.createElement(_antd.Button, {
    type: "primary",
    loading: submitting,
    onClick: handleSubmit,
    htmlType: "submit",
    className: "login-form-button"
  }, _t('Login')), auth.can_signup && /*#__PURE__*/_react.default.createElement("div", null, _t('Not registed'), "? ", /*#__PURE__*/_react.default.createElement("a", {
    onClick: () => nav('/signup')
  }, _t('please signup')))))));
};

var _default = SigininLayout;
exports.default = _default;
