// xadmin-antd@3.3.2/lib/auth/components/Signup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _xadmin = _interopRequireDefault(require("xadmin"));

var _antd = require("antd");

var _xadminI18n = require("xadmin-i18n");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = (_ref) => {
  var {
    error,
    children,
    invalid,
    handleSubmit,
    submitting
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_antd.Form, {
    className: "sigin-form",
    onSubmit: handleSubmit
  }, /*#__PURE__*/_react.default.createElement(_antd.Card, {
    style: {
      maxWidth: 450,
      margin: '5rem auto',
      marginBottom: '-0.5rem'
    },
    title: (0, _xadminI18n._t)('Please Signup')
  }, children, error && /*#__PURE__*/_react.default.createElement(_antd.Alert, {
    message: error,
    type: "error"
  }), /*#__PURE__*/_react.default.createElement(_antd.Form.Item, {
    wrapperCol: {
      xs: {
        span: 24
      },
      sm: {
        span: 18,
        offset: 5
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_antd.Button, {
    type: "primary",
    disabled: invalid,
    loading: submitting,
    onClick: handleSubmit,
    htmlType: "submit",
    className: "signup-form-button"
  }, (0, _xadminI18n._t)('Signup')), /*#__PURE__*/_react.default.createElement("div", null, (0, _xadminI18n._t)('Have account'), "? ", /*#__PURE__*/_react.default.createElement("a", {
    onClick: () => _xadmin.default.go('/login')
  }, (0, _xadminI18n._t)('please login')))))));
};

exports.default = _default;
