// xadmin-antd@3.3.2/es/auth/components/CaptchaCodeInput.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React from 'react';
import { Input, Col, Row, Tooltip } from 'antd';
import { app, api } from 'xadmin';
export default class CaptchaCodeInput extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      url: this.getCodeUrl()
    };
  }

  getCodeUrl() {
    var {
      field
    } = this.props;
    return api({}).host + (field.captcha_url || '/get_captcha_code') + '?random=' + Math.random().toString();
  }

  render() {
    var {
      input,
      field
    } = this.props;
    var {
      _t
    } = app.context;
    return /*#__PURE__*/React.createElement(Row, {
      gutter: 8
    }, /*#__PURE__*/React.createElement(Col, {
      span: 14
    }, /*#__PURE__*/React.createElement(Input, _extends({}, input, field.attrs))), /*#__PURE__*/React.createElement(Col, {
      span: 10
    }, /*#__PURE__*/React.createElement(Tooltip, {
      title: _t('Click to refresh captcha code')
    }, /*#__PURE__*/React.createElement("img", {
      style: {
        cursor: 'pointer',
        width: '100%'
      },
      onClick: () => this.setState({
        url: this.getCodeUrl()
      }),
      src: this.state.url
    }))));
  }

}
