// xadmin-antd@3.3.2/es/auth/components/Signin.js
import React from 'react';
import app, { use } from 'xadmin';
import { Form, Card, Button, Alert } from 'antd';

var SigininLayout = (_ref) => {
  var {
    error,
    children,
    invalid,
    handleSubmit,
    submitting
  } = _ref;
  var {
    _t
  } = app.context;
  var {
    auth
  } = app.get('config');
  var nav = use('navigate');
  return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(Form, {
    className: "sigin-form",
    onSubmit: handleSubmit
  }, /*#__PURE__*/React.createElement(Card, {
    style: {
      maxWidth: 450,
      margin: '5rem auto',
      marginBottom: '-0.5rem'
    },
    title: _t('Please Login')
  }, children, error && /*#__PURE__*/React.createElement(Alert, {
    message: error,
    type: "error"
  }), /*#__PURE__*/React.createElement(Form.Item, {
    wrapperCol: {
      xs: {
        span: 24
      },
      sm: {
        span: 18,
        offset: 5
      }
    }
  }, auth.can_reset_password && /*#__PURE__*/React.createElement("div", null, _t('Forgot password'), "? ", /*#__PURE__*/React.createElement("a", {
    onClick: () => nav('/forget_password')
  }, _t('reset password'))), /*#__PURE__*/React.createElement(Button, {
    type: "primary",
    loading: submitting,
    onClick: handleSubmit,
    htmlType: "submit",
    className: "login-form-button"
  }, _t('Login')), auth.can_signup && /*#__PURE__*/React.createElement("div", null, _t('Not registed'), "? ", /*#__PURE__*/React.createElement("a", {
    onClick: () => nav('/signup')
  }, _t('please signup')))))));
};

export default SigininLayout;
