// xadmin-antd@3.3.2/es/auth/components/Signup.js
import React from 'react';
import app from 'xadmin';
import { Form, Card, Button, Alert } from 'antd';
import { _t } from 'xadmin-i18n';
export default ((_ref) => {
  var {
    error,
    children,
    invalid,
    handleSubmit,
    submitting
  } = _ref;
  return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(Form, {
    className: "sigin-form",
    onSubmit: handleSubmit
  }, /*#__PURE__*/React.createElement(Card, {
    style: {
      maxWidth: 450,
      margin: '5rem auto',
      marginBottom: '-0.5rem'
    },
    title: _t('Please Signup')
  }, children, error && /*#__PURE__*/React.createElement(Alert, {
    message: error,
    type: "error"
  }), /*#__PURE__*/React.createElement(Form.Item, {
    wrapperCol: {
      xs: {
        span: 24
      },
      sm: {
        span: 18,
        offset: 5
      }
    }
  }, /*#__PURE__*/React.createElement(Button, {
    type: "primary",
    disabled: invalid,
    loading: submitting,
    onClick: handleSubmit,
    htmlType: "submit",
    className: "signup-form-button"
  }, _t('Signup')), /*#__PURE__*/React.createElement("div", null, _t('Have account'), "? ", /*#__PURE__*/React.createElement("a", {
    onClick: () => app.go('/login')
  }, _t('please login')))))));
});
