// xadmin-antd@3.3.2/lib/filter/filters.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FilterModal = exports.Submenu = exports.NavForm = exports.FilterForm = void 0;

var _react = _interopRequireDefault(require("react"));

var _xadmin = require("xadmin");

var _icons = require("@ant-design/icons");

var _antd = require("antd");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var {
  useBreakpoint
} = _antd.Grid;

var FilterForm = (_ref) => {
  var {
    children,
    invalid,
    handleSubmit,
    submitting,
    options,
    resetFilter
  } = _ref;
  var {
    _t
  } = _xadmin.app.context;
  return /*#__PURE__*/_react.default.createElement(_antd.Form, {
    onSubmit: handleSubmit
  }, children, options && options.submitOnChange ? null : /*#__PURE__*/_react.default.createElement(_antd.Form.Item, {
    style: {
      textAlign: 'center'
    }
  }, /*#__PURE__*/_react.default.createElement(_antd.Space, null, /*#__PURE__*/_react.default.createElement(_antd.Button, {
    disabled: invalid,
    loading: submitting,
    type: "primary",
    onClick: handleSubmit,
    icon: /*#__PURE__*/_react.default.createElement(_icons.SearchOutlined, null)
  }, _t('Search')), /*#__PURE__*/_react.default.createElement(_antd.Button, {
    disabled: submitting,
    onClick: resetFilter
  }, _t('Reset')))));
};

exports.FilterForm = FilterForm;

var NavForm = (_ref2) => {
  var {
    children,
    invalid,
    handleSubmit,
    submitting,
    options,
    resetFilter
  } = _ref2;
  var {
    _t
  } = _xadmin.app.context;
  return /*#__PURE__*/_react.default.createElement(_antd.Form, {
    layout: "inline",
    onSubmit: handleSubmit
  }, children, options && options.submitOnChange ? null : /*#__PURE__*/_react.default.createElement(_antd.Form.Item, null, /*#__PURE__*/_react.default.createElement(_antd.Space, null, /*#__PURE__*/_react.default.createElement(_antd.Button, {
    disabled: invalid,
    loading: submitting,
    type: "primary",
    onClick: handleSubmit,
    icon: /*#__PURE__*/_react.default.createElement(_icons.SearchOutlined, null)
  }, _t('Search')), /*#__PURE__*/_react.default.createElement(_antd.Button, {
    disabled: submitting,
    onClick: resetFilter
  }, _t('Reset')))));
};

exports.NavForm = NavForm;

var FilterOpenLink = (_ref3) => {
  var {
    count,
    onClick,
    show
  } = _ref3;
  var {
    _t
  } = _xadmin.app.context;
  var screens = useBreakpoint();
  return screens.xxl == false && count > 3 || count > 4 ? /*#__PURE__*/_react.default.createElement(_antd.Typography.Link, {
    onClick: onClick
  }, show ? _t('Collapse') : _t('Expand')) : null;
};

var Submenu = (_ref4) => {
  var {
    children,
    invalid,
    handleSubmit,
    submitting,
    options,
    resetFilter
  } = _ref4;
  var {
    _t
  } = _xadmin.app.context;
  var defaultShowAllFilter = (0, _xadmin.config)('filter') && (0, _xadmin.config)('filter').submenuShowAllFilter == true;

  var [showAllFilter, setShowAllFilter] = _react.default.useState(defaultShowAllFilter);

  return /*#__PURE__*/_react.default.createElement(_antd.Form, {
    className: "ant-advanced-search-form",
    onSubmit: handleSubmit
  }, /*#__PURE__*/_react.default.createElement(_antd.Card, {
    style: {
      marginBottom: '.5rem',
      overflow: 'hidden'
    }
  }, /*#__PURE__*/_react.default.createElement(_antd.Row, {
    gutter: 8,
    style: {
      flexWrap: children.length <= 3 || showAllFilter ? 'wrap' : 'nowrap'
    }
  }, children), options && options.submitOnChange ? null : /*#__PURE__*/_react.default.createElement(_antd.Row, null, /*#__PURE__*/_react.default.createElement(_antd.Col, {
    span: 24,
    style: {
      textAlign: 'center'
    }
  }, /*#__PURE__*/_react.default.createElement(_antd.Space, null, /*#__PURE__*/_react.default.createElement(_antd.Button, {
    disabled: invalid,
    loading: submitting,
    type: "primary",
    onClick: handleSubmit,
    icon: /*#__PURE__*/_react.default.createElement(_icons.SearchOutlined, null)
  }, _t('Search')), /*#__PURE__*/_react.default.createElement(_antd.Button, {
    disabled: submitting,
    onClick: resetFilter
  }, _t('Reset')), /*#__PURE__*/_react.default.createElement(FilterOpenLink, {
    count: children.length,
    onClick: () => setShowAllFilter(!showAllFilter),
    show: showAllFilter
  }))))));
};

exports.Submenu = Submenu;

class FilterModal extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      show: false
    });

    _defineProperty(this, "onClose", () => {
      this.setState({
        show: false
      });
    });
  }

  render() {
    var {
      _t
    } = _xadmin.app.context;
    var {
      children,
      invalid,
      handleSubmit,
      submitting,
      options,
      resetFilter
    } = this.props;
    var icon = submitting ? 'spinner fa-spin' : 'floppy-o';
    var buttons = options && options.submitOnChange ? {} : {
      okText: _t('Search'),
      cancelText: _t('Reset'),
      okButtonProps: {
        disabled: invalid,
        loading: submitting,
        icon: /*#__PURE__*/_react.default.createElement(_icons.SearchOutlined, null)
      },
      onOk: () => {
        handleSubmit();
        this.onClose();
      },
      onCancel: () => {
        resetFilter();
        this.onClose();
      }
    };
    return [/*#__PURE__*/_react.default.createElement(_antd.Button, {
      key: "filter-btn",
      onClick: () => this.setState({
        show: true
      })
    }, /*#__PURE__*/_react.default.createElement(_icons.FilterOutlined, null), " ", _t('Filter')), /*#__PURE__*/_react.default.createElement(_antd.Modal, _extends({
      key: "filter-modal",
      title: _t('Filter Form'),
      visible: this.state.show
    }, buttons), /*#__PURE__*/_react.default.createElement(_antd.Form, {
      onSubmit: handleSubmit
    }, children))];
  }

}

exports.FilterModal = FilterModal;
