// @gtiot/iot-devtool@1.4.5/config/webpack.config.js
var path = require("path");
const { execFileSync, exec } = require('child_process');
const paths = require('./paths');

const fs = require('fs');
const HtmlWebPackPlugin = require("html-webpack-plugin");
const HtmlWebpackTagsPlugin = require('html-webpack-tags-plugin');
const BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin;
const { CleanWebpackPlugin } = require('clean-webpack-plugin');
const CreateFilePlugin = require('./createFile')

const webpack = require('webpack');
const { isLocal, getModule } = require("../utils/envs");
const extConfig = require('./externals.json')

const isCI = process.env.CI &&
(typeof process.env.CI !== "string" ||
  process.env.CI.toLowerCase() !== "false")

const appPackage = require(paths.appPackageJson);

let appWebpackConfig;
try {
  appWebpackConfig = require(paths.appWebpackConfig)
} catch (err) {
  if(err.code != 'MODULE_NOT_FOUND') {
    throw err;
  }
}

const gitInfo = (...args) => {
  return execFileSync('git', args, {
    cwd: paths.appPath,
    encoding: 'utf8',
    timeout: 7000,
  }).replace(/[^\da-z]*/gim, '')
}

function dateFtt(fmt, date) { //author: meizz 
  var o = {
    "M+": date.getMonth() + 1,     //月份 
    "d+": date.getDate(),     //日 
    "h+": date.getHours(),     //小时 
    "m+": date.getMinutes(),     //分 
    "s+": date.getSeconds(),     //秒 
    "q+": Math.floor((date.getMonth() + 3) / 3), //季度 
    "S": date.getMilliseconds()    //毫秒 
  };
  if (/(y+)/.test(fmt))
    fmt = fmt.replace(RegExp.$1, (date.getFullYear() + "").substr(4 - RegExp.$1.length));
  for (var k in o)
    if (new RegExp("(" + k + ")").test(fmt))
      fmt = fmt.replace(RegExp.$1, (RegExp.$1.length == 1) ? (o[k]) : (("00" + o[k]).substr(("" + o[k]).length)));
  return fmt;
}

module.exports = async (env, host) => {
  const isEnvProduction = env === 'production';
  const useLocal = isLocal()
  const modules = getModule(env)

  const defaultModules = modules == 'front' ? [
    '@gtiot/iot-client/dist/front.js',
    '@gtiot/iot-theme/dist/front.js',
    '@gtiot/iot-dashboard/dist/front.js'
  ] : [
    '@gtiot/iot-client/dist/index.js',
    '@gtiot/iot-theme/dist/index.js',
    '@gtiot/iot-dashboard/dist/index.js'
  ]

  let iotModulesScripts = []

  // 导入 externals 包
  if(useLocal) {
    iotModulesScripts = [
      ...(
        isEnvProduction ? extConfig.externalsScripts : extConfig.externalsScriptsDev
      )
    ]
  } else {
    iotModulesScripts = [
      isEnvProduction ? '@gtiot/iot-dll/dist/iot.min.js' : '@gtiot/iot-dll/dist/iot.js',
    ]
  }

  // 导入默认包
  defaultModules.forEach(modeulePath => {
    if(modeulePath.indexOf(appPackage.name) != 0 && iotModulesScripts.indexOf(modeulePath) == -1) {
      iotModulesScripts.push(modeulePath);
    }
  })

  // 导入依赖iot包
  if(appPackage.iotDependencies) {
    appPackage.iotDependencies.forEach(modeulePath => {
      if(iotModulesScripts.indexOf(modeulePath) == -1) {
        const filePath = modeulePath.endsWith('.js') ? modeulePath : modeulePath + (modules == 'front' ? '/dist/front.js' : '/dist/index.js')
        iotModulesScripts.push(filePath);
      }
    })
  }

  // 转换成在线或本地模式
  iotModulesScripts = iotModulesScripts.map(url => (useLocal ? 'node_modules/' : (host + '/node_modules/')) + url)

  const entry = {}
  if(modules == 'admin' || modules == 'all') {
    entry['index'] = paths.appIndexJs
  }
  if((modules == 'front' || modules == 'all') && fs.existsSync(paths.appFrontJs) ) {
    entry['front'] = paths.appFrontJs
  }

  appPackage.dependencies = {}

  if(!isCI) {
    try {
      gitInfo('log')
      const branch = gitInfo('rev-parse', '--abbrev-ref', 'HEAD')
      const longHead = gitInfo('rev-parse', 'HEAD')
      appPackage.gitHead = longHead
    
      if(branch != 'master' && branch != 'HEAD') {
        const gitHead = gitInfo('rev-parse', '--short', 'HEAD')
        appPackage.version =  appPackage.version + '+' + branch + '.' + gitHead
    
        appPackage.buildTime = dateFtt('yyyy-MM-dd hh:mm:ss', new Date())
        appPackage.buildUser = gitInfo('config', 'user.name')
      }
    } catch (error) {
      
    }
  }

  let config = {
    mode: env,
    entry,
    output: {
      path: paths.buildPath,
      filename: "[name].js",
      chunkFilename: "chunks/[name].[chunkhash].js",
      publicPath: process.env.NODE_ENV === "production" ? '/node_modules/' + appPackage.name + '/dist/' : '/',
      globalObject: 'this'
    },
    resolve: {
      alias: {
        '@': paths.appSrc
      },
      modules: ['node_modules'].concat(
        process.env.NODE_PATH.split(path.delimiter).filter(Boolean)
      ),
      fallback: {
        path: require.resolve("path-browserify")
      }
    },
    externals: extConfig.externals,
    module: {
      rules: [
        {
          test: /\.(js|jsx)$/,
          exclude: /node_modules/,
          use: {
            loader: require.resolve("babel-loader"),
            options: {
              presets: [
                require.resolve("@babel/preset-env"),
                require.resolve("@babel/preset-react")
              ],
              plugins: [
                require.resolve("@babel/plugin-transform-runtime"),
                require.resolve("@babel/plugin-proposal-object-rest-spread"),
                [require.resolve("@babel/plugin-proposal-decorators"), { "legacy": true }],
                [require.resolve("@babel/plugin-proposal-class-properties"), { "loose": true }]
              ]
            }
          }
        },
        {
          test: /\.css$/,
          use: [
            require.resolve('style-loader'),
            require.resolve('css-loader')
          ]
        },
        {
          test: /^((?!theme).)+\.less$/,
          use: [
            require.resolve('style-loader'),
            require.resolve('css-loader'),
            {
              loader: require.resolve('less-loader'),
              options: {
                lessOptions: {
                  javascriptEnabled: true,
                }
              }
            }
          ]
        },
        {
          test: /\.theme\.less$/,
          use: [
            {
              loader: require.resolve('file-loader'),
              options: {
                name: '[hash].theme.css',
                publicPath: process.env.NODE_ENV === "production" ? '/node_modules/' + appPackage.name + '/dist' : undefined,
                esModule: false,
              },
            },
            {
              loader: require.resolve('less-loader'),
              options: {
                lessOptions: {
                  javascriptEnabled: true,
                }
              }
            }
          ]
        },
        {
          test: /\.svg$/,
          oneOf: [
            {
              resourceQuery: /inline/,
              loader: require.resolve("url-loader"),
              options: {
                esModule: false,
              }
            },
            {
              loader: require.resolve('svg-inline-loader')
            }
          ]
        },
        {
          test: /\.html$/,
          use: [
            {
              loader: require.resolve("html-loader")
            }
          ]
        },
        {
          test: /\.(jpg|png|gif|ttf)/,
          use: [
            {
              loader: require.resolve("file-loader"),
              options: {
                esModule: false,
              }
            }
          ]
        }
      ]
    },
    plugins: isEnvProduction ? [
      new CleanWebpackPlugin(),
      !isCI && new CreateFilePlugin({
        path: paths.buildPath,
        fileName: '.buildinfo.json',
        content: JSON.stringify(appPackage, 0, 2)
      }),
      process.env.WITH_ANALYZER && new BundleAnalyzerPlugin()
    ].filter(Boolean) : [
      new HtmlWebPackPlugin({
        template: paths.appHtml,
        filename: "./index.html"
      }),
      new HtmlWebpackTagsPlugin({ scripts: iotModulesScripts, append: false, publicPath: useLocal }),
      new HtmlWebpackTagsPlugin({ scripts: ['start.js'], append: true }),
      new webpack.HotModuleReplacementPlugin(),
    ].filter(Boolean),
    devtool: isEnvProduction ? 'nosources-source-map' : 'eval-cheap-module-source-map',
    performance: {
      hints: false
    }
  }
  if(appWebpackConfig && typeof appWebpackConfig === 'function') {
    config = appWebpackConfig(config)
  }
  return config
};
