// xadmin-antd@3.3.2/lib/filter/components/DateFilter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var _moment = _interopRequireDefault(require("moment"));

var _lodash = _interopRequireDefault(require("lodash"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var {
  MonthPicker,
  RangePicker,
  WeekPicker
} = _antd.DatePicker;

class DatePickerFilter extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "onRangeChange", value => {
      var {
        input
      } = this.props;
      input.onChange(this.getValue(value));
    });
  }

  getValue(value) {
    var {
      field
    } = this.props;
    var format = field.datetimeFormat || 'YYYY-MM-DD';
    return {
      rule: 'range',
      gte: value[0].format(format),
      lte: value[1].format(format)
    };
  }

  render() {
    var {
      input,
      field
    } = this.props;
    var value = input.value;

    if (_lodash.default.isPlainObject(value) && value.gte && value.lte) {
      var format = field.datetimeFormat || 'YYYY-MM-DD';
      value = [(0, _moment.default)(value.gte, format), (0, _moment.default)(value.lte, format)];
    }

    return /*#__PURE__*/_react.default.createElement(RangePicker, {
      onChange: this.onRangeChange,
      value: value
    });
  }

}

exports.default = DatePickerFilter;
