// xadmin-ui@3.3.2/dist/xadmin-ui.js
(function (global, factory) {
typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('react'), require('xadmin'), require('react-router-dom')) :
typeof define === 'function' && define.amd ? define(['exports', 'react', 'xadmin', 'react-router-dom'], factory) :
(global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.xadminUi = {}, global.React, global.xadmin, global.ReactRouterDOM));
}(this, (function (exports, React, app, reactRouterDom) { 'use strict';

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var app__default = /*#__PURE__*/_interopDefaultLegacy(app);

function unwrapExports (x) {
	return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
}

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var _extends_1 = createCommonjsModule(function (module) {
function _extends() {
  module.exports = _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  module.exports["default"] = module.exports, module.exports.__esModule = true;
  return _extends.apply(this, arguments);
}

module.exports = _extends;
module.exports["default"] = module.exports, module.exports.__esModule = true;
});

var _extends = unwrapExports(_extends_1);

var defineProperty = createCommonjsModule(function (module) {
function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

module.exports = _defineProperty;
module.exports["default"] = module.exports, module.exports.__esModule = true;
});

var _defineProperty = unwrapExports(defineProperty);

var objectWithoutPropertiesLoose = createCommonjsModule(function (module) {
function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

module.exports = _objectWithoutPropertiesLoose;
module.exports["default"] = module.exports, module.exports.__esModule = true;
});

unwrapExports(objectWithoutPropertiesLoose);

var objectWithoutProperties = createCommonjsModule(function (module) {
function _objectWithoutProperties(source, excluded) {
  if (source == null) return {};
  var target = objectWithoutPropertiesLoose(source, excluded);
  var key, i;

  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

    for (i = 0; i < sourceSymbolKeys.length; i++) {
      key = sourceSymbolKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
      target[key] = source[key];
    }
  }

  return target;
}

module.exports = _objectWithoutProperties;
module.exports["default"] = module.exports, module.exports.__esModule = true;
});

var _objectWithoutProperties = unwrapExports(objectWithoutProperties);

var splashscreen = {
  start: function start(app) {
    return function () {
      var loaderDiv = document.getElementById('loader-wrapper');

      if (loaderDiv) {
        loaderDiv.addEventListener('transitionend', function () {
          loaderDiv.style.display = 'none';
        }, true);
        loaderDiv.style.display = 'none';
      }
    };
  }
};

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var C = function C(args) {
  if (typeof args == 'string') {
    return app__default['default'].get('components')[args];
  } else {
    var is = args.is,
        props = _objectWithoutProperties(args, ["is"]);

    var Component = C(is);
    if (Component) return /*#__PURE__*/React__default['default'].createElement(Component, props);else return /*#__PURE__*/React__default['default'].createElement("div", null, "Component ", is, " not found.");
  }
};

C.lazy = function (is) {
  return function (props) {
    return C(_objectSpread({
      is: is
    }, props));
  };
};

var Main = C.lazy('Main');
var App = C.lazy('App');
var Dashboard = C.lazy('Dashboard');
var Page = C.lazy('Page');
var Icon = C.lazy('Icon');
var Loading = C.lazy('Loading');
var Dropdown = C.lazy('Dropdown');
var Menu = C.lazy('Menu');
Menu.SubMenu = C.lazy('Menu.SubMenu');
Menu.Item = C.lazy('Menu.Item');
/** UI Components */

var Badge = C.lazy('Badge');
var Card = C.lazy('Card');
var Modal = C.lazy('Modal');
var Button = C.lazy('Button');
var Popover = C.lazy('Popover');
var Tooltip = C.lazy('Tooltip');
var Table = C.lazy('Table');
var Tabs = C.lazy('Tabs');
Tabs.Item = C.lazy('Tabs.Item');
var Empty = C.lazy('Empty');
var List = C.lazy('List');
List.Item = C.lazy('List.Item');
var Alert = C.lazy('Alert');
var Input = C.lazy('Input');
Input.Static = C.lazy('Input.Static');
var Check = C.lazy('Check');
var Select = C.lazy('Select');

var lazy = function lazy(fn, fallback) {
  var Component = /*#__PURE__*/React__default['default'].lazy(fn);
  return /*#__PURE__*/React__default['default'].forwardRef(function (props, ref) {
    return /*#__PURE__*/React__default['default'].createElement(React__default['default'].Suspense, {
      fallback: fallback || /*#__PURE__*/React__default['default'].createElement(Loading, null)
    }, /*#__PURE__*/React__default['default'].createElement(Component, _extends({
      ref: ref
    }, props)));
  });
};

var index = {
  name: 'xadmin.ui',
  items: {
    components: {
      type: 'map'
    }
  },
  routers: {
    '@': {
      path: '/',
      component: Main
    },
    '/': [{
      index: true,
      element: /*#__PURE__*/React__default['default'].createElement(reactRouterDom.Navigate, {
        to: "/app/",
        replace: true
      })
    }, {
      path: 'app/',
      component: App
    }],
    '/app/': [{
      index: true,
      component: Dashboard
    }, {
      path: 'dashboard',
      component: Dashboard
    }]
  },
  reducers: {
    loading: function loading() {
      var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      var _ref = arguments.length > 1 ? arguments[1] : undefined,
          type = _ref.type,
          key = _ref.key;

      if (!key) {
        return state;
      }

      switch (type) {
        case 'START_LOADING':
          return _objectSpread(_objectSpread({}, state), {}, _defineProperty({}, key, true));

        case 'END_LOADING':
          return _objectSpread({}, _.omit(state, key));

        default:
          return state;
      }
    }
  }
};
var apps = {
  splashscreen: splashscreen
};

exports.Alert = Alert;
exports.App = App;
exports.Badge = Badge;
exports.Button = Button;
exports.C = C;
exports.Card = Card;
exports.Check = Check;
exports.Dashboard = Dashboard;
exports.Dropdown = Dropdown;
exports.Empty = Empty;
exports.Icon = Icon;
exports.Input = Input;
exports.List = List;
exports.Loading = Loading;
exports.Main = Main;
exports.Menu = Menu;
exports.Modal = Modal;
exports.Page = Page;
exports.Popover = Popover;
exports.Select = Select;
exports.Table = Table;
exports.Tabs = Tabs;
exports.Tooltip = Tooltip;
exports.apps = apps;
exports.default = index;
exports.lazy = lazy;

Object.defineProperty(exports, '__esModule', { value: true });

})));
