// xadmin-antd@3.3.2/es/layout/Page.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import { PageHeader } from 'antd';
import { Layout, Menu } from 'antd';
import React from 'react';

class Page extends React.Component {
  render() {
    var _this$props = this.props,
        {
      title,
      subTitle,
      content,
      breadcrumb,
      extra,
      actions,
      tabs,
      footer,
      className,
      style,
      children
    } = _this$props,
        pageProps = _objectWithoutProperties(_this$props, ["title", "subTitle", "content", "breadcrumb", "extra", "actions", "tabs", "footer", "className", "style", "children"]);

    return /*#__PURE__*/React.createElement("div", {
      className: className,
      style: style
    }, /*#__PURE__*/React.createElement(PageHeader, _extends({
      title: title,
      subTitle: subTitle,
      extra: actions,
      tabs: tabs,
      breadcrumb: breadcrumb,
      footer: footer,
      style: {
        margin: 10
      }
    }, pageProps), content || extra ? /*#__PURE__*/React.createElement(React.Fragment, null, content && /*#__PURE__*/React.createElement("div", {
      className: "content padding"
    }, content), extra && /*#__PURE__*/React.createElement("div", {
      className: "extraContent"
    }, extra)) : null), /*#__PURE__*/React.createElement("div", {
      style: {
        margin: 10
      }
    }, children));
  }

}

export default Page;
