// xadmin-i18n@3.1.0/es/index.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import _ from 'lodash';
import moment from 'moment';
import i18next from 'i18next';
import XHR from 'i18next-xhr-backend';
import Cache from 'i18next-localstorage-cache';
import LanguageDetector from 'i18next-browser-languagedetector';
import app from 'xadmin';
import zh_Hans from './zh_Hans/translation.json';
var locales = {
  zh_Hans: {
    translation: zh_Hans
  }
};

var _t = function _t() {
  return app.context._t(...arguments);
};

export { _t, i18next };
export default {
  name: 'xadmin.i18n',
  items: {
    locales: {
      type: 'mapArray'
    }
  },
  context: app => (context, cb) => {
    var resources = app.get('locales');
    var {
      locale
    } = app.get('config');

    for (var ln in resources) {
      resources[ln] = _.merge({}, ...resources[ln]);
    }

    i18next.use(XHR) // or any other backend implementation
    .use(Cache) // or any other cache implementation
    .use(LanguageDetector) // or any other implementation
    .init(_objectSpread({
      debug: false,
      lng: 'en',
      fallbackLng: false,
      keySeparator: false,
      nsSeparator: false,
      resources
    }, locale || {}), (err, t) => {
      moment.locale(locale && locale.moment || 'en');
      cb(null, _objectSpread(_objectSpread({}, context), {}, {
        _t: t,
        i18n: i18next
      }));
    });
  },
  locales
};
