// xadmin-antd@3.3.2/es/filter/components/DateFilter.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import { DatePicker } from 'antd';
import moment from 'moment';
import _ from 'lodash';
var {
  MonthPicker,
  RangePicker,
  WeekPicker
} = DatePicker;
export default class DatePickerFilter extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "onRangeChange", value => {
      var {
        input
      } = this.props;
      input.onChange(this.getValue(value));
    });
  }

  getValue(value) {
    var {
      field
    } = this.props;
    var format = field.datetimeFormat || 'YYYY-MM-DD';
    return {
      rule: 'range',
      gte: value[0].format(format),
      lte: value[1].format(format)
    };
  }

  render() {
    var {
      input,
      field
    } = this.props;
    var value = input.value;

    if (_.isPlainObject(value) && value.gte && value.lte) {
      var format = field.datetimeFormat || 'YYYY-MM-DD';
      value = [moment(value.gte, format), moment(value.lte, format)];
    }

    return /*#__PURE__*/React.createElement(RangePicker, {
      onChange: this.onRangeChange,
      value: value
    });
  }

}
