// xadmin-antd@3.3.2/es/filter/components/EnumFilter.js
import React from 'react';
import _ from 'lodash';
import { Select } from 'antd';

var FilterEnum = props => {
  var {
    input: {
      value,
      onChange
    },
    field
  } = props;
  var placeholder = field && field.placeholder;
  var titleMap = field && field.titleMap;

  var selectChange = values => {
    var v = null;

    if (values.length > 1) {
      v = {
        $in: values
      };
    } else if (values.length > 0) {
      v = values[0];
    }

    onChange(v);
  };

  var seleted = []; // 初始化value

  if (value && value != '') {
    if (_.isString(value)) seleted = [value];
    if (_.isObject(value)) seleted = value.$in;
  }

  return /*#__PURE__*/React.createElement(Select, {
    mode: "multiple",
    style: {
      width: '100%'
    },
    placeholder: placeholder ? placeholder : '请选择',
    value: seleted,
    onChange: selectChange
  }, titleMap && titleMap.length > 0 && titleMap.map(item => {
    return /*#__PURE__*/React.createElement(Select.Option, {
      key: item.value
    }, item.name);
  }));
};

export default FilterEnum;
