// xadmin-antd@3.3.2/es/model/components/ChildrenModel.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import _ from 'lodash';
import app, { use } from 'xadmin';
import { _t } from 'xadmin-i18n';
import { Button, Modal, Form } from 'antd';
import { SchemaForm } from 'xadmin-form';
import { Model } from 'xadmin-model';
import { C, Icon } from 'xadmin-ui';

var ChildrenModel = props => {
  var [show, setShow] = React.useState(false);

  var {
    parent,
    model,
    refFilter,
    refData,
    refField,
    modelProps,
    children,
    header,
    value,
    onClose,
    refreshTimeout
  } = props,
      extProps = _objectWithoutProperties(props, ["parent", "model", "refFilter", "refData", "refField", "modelProps", "children", "header", "value", "onClose", "refreshTimeout"]);

  var handleCancel = () => {
    setShow(false);
    onClose && onClose();
  };

  var cmodel = _.isString(model) ? app.get('models')[model] : model;

  var schema = _objectSpread(_objectSpread({}, cmodel), {}, {
    parent,
    itemActions: [...(cmodel.itemActions || []), item => /*#__PURE__*/React.createElement(EditChildrenModelBtn, {
      id: item.id,
      refData: refData,
      refreshTimeout: refreshTimeout
    }, _t('Edit'))],
    permission: _objectSpread(_objectSpread({}, cmodel.permission), {}, {
      edit: false,
      childEdit: cmodel.permission && cmodel.permission.edit
    })
  }, modelProps);

  var initialValues = {
    wheres: {
      filters: refFilter || {
        [refField]: parent.id
      }
    }
  };
  var action = children && !_.isString(children) && /*#__PURE__*/React.isValidElement(children) ? /*#__PURE__*/React.cloneElement(children, {
    onClick: () => setShow(true)
  }) : /*#__PURE__*/React.createElement(Button, _extends({
    size: "small",
    key: "child-model-action",
    className: "model-list-action"
  }, extProps, {
    onClick: () => setShow(true)
  }), children || cmodel.title || cmodel.name);
  var ItemsComponent = cmodel.components && cmodel.components.DataList || C('Model.DataTable');
  return [action, show ? /*#__PURE__*/React.createElement(Model, {
    schema: schema,
    modelKey: "".concat(_.isString(model) ? model : model.name, "_").concat(parent.id),
    initialValues: initialValues
  }, /*#__PURE__*/React.createElement(Modal, {
    key: 1,
    visible: show,
    title: header || cmodel.title || cmodel.name,
    width: "80%",
    onCancel: handleCancel,
    onOk: handleCancel
  }, /*#__PURE__*/React.createElement("div", {
    key: "model-list-subnav",
    style: {
      display: 'flex',
      justifyContent: 'space-between',
      marginBottom: '.5rem'
    }
  }, /*#__PURE__*/React.createElement(C, {
    is: "Model.Pagination"
  }), /*#__PURE__*/React.createElement(C, {
    is: "Model.ListSubMenu"
  }, /*#__PURE__*/React.createElement(AddChildrenModelBtn, props))), /*#__PURE__*/React.createElement(ItemsComponent, {
    key: "model-list-grid"
  }), /*#__PURE__*/React.createElement("div", {
    key: "model-list-downnav",
    style: {
      display: 'flex',
      justifyContent: 'space-between',
      marginTop: '.5rem'
    }
  }, /*#__PURE__*/React.createElement(C, {
    is: "Model.ActionBar"
  }), /*#__PURE__*/React.createElement(C, {
    is: "Model.Pagination"
  })))) : null];
};

var AddChildrenModelBtn = props => {
  var [show, setShow] = React.useState(false);
  var {
    model
  } = use('model');
  var {
    getItems
  } = use('model.getItems');
  var {
    canAdd
  } = use('model.permission');
  var {
    refData,
    refreshTimeout
  } = props;

  var onSuccess = () => {
    if (refreshTimeout) {
      setTimeout(getItems, refreshTimeout);
    } else {
      getItems();
    }
  };

  return canAdd ? [/*#__PURE__*/React.createElement(Button, {
    key: 0,
    style: {
      marginLeft: '.5rem'
    },
    onClick: () => setShow(true)
  }, _t('Add {{object}}', {
    object: model.title
  })), /*#__PURE__*/React.createElement(ChildrenFormModel, {
    key: 1,
    onSuccess: onSuccess,
    refData: refData,
    show: show,
    onClose: () => setShow(false)
  })] : null;
};

var EditChildrenModelBtn = props => {
  var [show, setShow] = React.useState(false);
  var {
    model
  } = use('model');
  var {
    getItems
  } = use('model.getItems');
  var {
    id,
    refData,
    refreshTimeout
  } = props;
  var canChildEdit = !!model.permission && !!model.permission.childEdit;

  var onSuccess = () => {
    if (refreshTimeout) {
      setTimeout(getItems, refreshTimeout);
    } else {
      getItems();
    }
  };

  return canChildEdit && [/*#__PURE__*/React.createElement(Button, {
    key: 0,
    size: "sm",
    className: "model-list-action",
    onClick: () => setShow(true)
  }, _t('Edit')), show ? /*#__PURE__*/React.createElement(ChildrenFormModel, {
    key: 1,
    onSuccess: onSuccess,
    id: id,
    refData: refData,
    show: show,
    onClose: () => {
      setShow(false);
    }
  }) : null];
};

var ChildrenFormModel = props => {
  var {
    show,
    title,
    onClose,
    modalProps,
    onSuccess,
    refData
  } = props;
  var {
    model
  } = use('model');
  var {
    data,
    loading,
    saveItem
  } = use('model.item', props);

  var onSubmitSuccess = item => {
    onClose();
    onSuccess(item);
  };

  var onSaveItem = values => {
    saveItem(_objectSpread(_objectSpread({}, values), refData));
  };

  var FormLayout = props => {
    var {
      children,
      invalid,
      handleSubmit,
      submitting
    } = props;
    return /*#__PURE__*/React.createElement(Modal, _extends({}, modalProps, {
      visible: show,
      width: "70%",
      title: title,
      className: "xadmin-modal-form",
      onCancel: onClose,
      onOk: handleSubmit,
      okButtonDisabled: invalid || submitting,
      okText: /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Icon, {
        name: "ban"
      }), " ", _t('Save')),
      okButtonProps: {
        loading: submitting
      }
    }), /*#__PURE__*/React.createElement(Form, null, children));
  };

  return show && !loading ? /*#__PURE__*/React.createElement(SchemaForm, {
    formKey: "model.modalform.".concat(model.key),
    schema: model,
    initialValues: _objectSpread(_objectSpread({}, data), refData),
    onSubmit: onSaveItem,
    onClose: onClose,
    component: FormLayout,
    onSubmitSuccess: onSubmitSuccess
  }) : null;
};

export default ChildrenModel;
