// xadmin-antd@3.3.2/es/model/components/DetailModal.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import _ from 'lodash';
import { Modal } from 'antd';
import { C } from 'xadmin-ui';

class DetailModal extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      show: false
    });
  }

  showModal() {
    this.setState({
      show: true
    });
  }

  hideModal() {
    this.setState({
      show: false
    });
  }

  render() {
    var {
      id,
      children
    } = this.props;
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("a", {
      style: {
        cursor: 'pointer'
      },
      onClick: this.showModal
    }, children), this.state.show && /*#__PURE__*/React.createElement(Modal, {
      visible: this.state.show,
      onCancel: this.hideModal,
      title: children
    }, /*#__PURE__*/React.createElement(C, {
      is: "Model.DataDetail",
      id: id
    })));
  }

}

export default DetailModal;
