// xadmin-antd@3.3.2/es/model/components/Pages.js
import { Button, Layout } from 'antd';
import React from 'react';
import _ from 'lodash';
import { use } from 'xadmin';
import { _t } from 'xadmin-i18n';
import { ModelBlock } from 'xadmin-model';
import { C, Icon, Page, Loading } from 'xadmin-ui';
var {
  Content,
  Sider
} = Layout;

var DefaultAddButton = (_ref) => {
  var {
    onAdd,
    children
  } = _ref;
  return /*#__PURE__*/React.createElement(Button, {
    type: "primary",
    onClick: onAdd
  }, children);
};

var ModelListPage = () => {
  var {
    model
  } = use('model');
  var {
    canAdd
  } = use('model.permission');
  var {
    onAdd
  } = use('model.event');
  var query = use('searchParams');
  var ItemsComponent = model.components && model.components.DataList || C('Model.DataTable');
  var icon = model.icon || model.name;

  if (_.isString(icon)) {
    icon = /*#__PURE__*/React.createElement(Icon, {
      name: icon
    });
  }

  var title = model.title;

  var renderActions = () => {
    var AddButton = model.components && model.components.AddButton || C('Model.AddButton') || DefaultAddButton;
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(ModelBlock, {
      name: "model.list.navbtn"
    }), canAdd ? /*#__PURE__*/React.createElement(AddButton, {
      onAdd: onAdd
    }, /*#__PURE__*/React.createElement(Icon, {
      name: "plus"
    }), " ", _t('Add {{object}}', {
      object: model.title
    })) : null);
  };

  var GridComponents = [/*#__PURE__*/React.createElement("div", {
    key: "model-list-subnav",
    style: {
      display: 'flex',
      justifyContent: 'space-between',
      marginBottom: '.5rem'
    }
  }, /*#__PURE__*/React.createElement(C, {
    is: "Model.Pagination"
  }), /*#__PURE__*/React.createElement(C, {
    is: "Model.ListSubMenu"
  })), /*#__PURE__*/React.createElement(ItemsComponent, {
    key: "model-list-grid",
    query: query
  }), /*#__PURE__*/React.createElement("div", {
    key: "model-list-downnav",
    style: {
      display: 'flex',
      justifyContent: 'space-between',
      marginTop: '.5rem'
    }
  }, /*#__PURE__*/React.createElement(C, {
    is: "Model.ActionBar"
  }), /*#__PURE__*/React.createElement(C, {
    is: "Model.Pagination"
  }))];
  return /*#__PURE__*/React.createElement(Page, {
    className: "xadmin-model-list-".concat(model.key),
    title: /*#__PURE__*/React.createElement("span", null, icon, " ", title),
    subTitle: /*#__PURE__*/React.createElement(ModelBlock, {
      name: "model.list.nav"
    }),
    actions: renderActions()
  }, /*#__PURE__*/React.createElement(ModelBlock, {
    name: "model.list.submenu"
  }), /*#__PURE__*/React.createElement(ModelBlock, {
    name: "model.list.sidemenu"
  }, sideMenu => /*#__PURE__*/React.createElement(ModelBlock, {
    name: "model.list.sidepanel"
  }, sidePanel => sideMenu || sidePanel ? /*#__PURE__*/React.createElement(Layout, null, sideMenu ? /*#__PURE__*/React.createElement(Sider, {
    width: 200,
    style: {
      backgroundColor: 'transparent',
      marginRight: '.5rem'
    }
  }, sideMenu) : null, /*#__PURE__*/React.createElement(Content, null, GridComponents), sidePanel ? /*#__PURE__*/React.createElement(Sider, {
    width: 200,
    style: {
      backgroundColor: 'transparent',
      marginLeft: '.5rem'
    }
  }, sidePanel) : null) : GridComponents)));
};

var ModelFormPage = () => {
  var {
    model
  } = use('model');
  var {
    id
  } = use('params');
  var {
    onSaved
  } = use('model.event');
  var {
    data,
    loading
  } = use('model.item', {
    id
  });
  var query = use('query');
  var title = id ? _t('Edit {{title}}', {
    title: model.title + ' ' + (data && data[model.displayField || 'name'] || '')
  }) : _t('Create {{title}}', {
    title: model.title
  });
  var FormComponent = model.components && model.components.DataForm || C('Model.DataForm');
  return /*#__PURE__*/React.createElement(Page, {
    className: "xadmin-model-form-".concat(model.key),
    title: title,
    onBack: () => history.back(),
    subTitle: /*#__PURE__*/React.createElement(ModelBlock, {
      name: "model.edit.nav"
    }),
    actions: /*#__PURE__*/React.createElement(ModelBlock, {
      name: "model.edit.navbtn"
    })
  }, /*#__PURE__*/React.createElement(ModelBlock, {
    name: "model.form.before"
  }), loading ? /*#__PURE__*/React.createElement(Loading, null) : /*#__PURE__*/React.createElement(FormComponent, {
    id: id,
    item: data,
    query: query,
    onSubmitSuccess: onSaved
  }), /*#__PURE__*/React.createElement(ModelBlock, {
    name: "model.form.after"
  }));
};

var ModelDetailPage = () => {
  var {
    model
  } = use('model');
  var {
    id
  } = use('params');

  var renderActions = () => {
    var {
      canEdit
    } = use('model.permission');
    var {
      onEdit
    } = use('model.event');
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(ModelBlock, {
      name: "model.detail.navbtn"
    }), canEdit ? /*#__PURE__*/React.createElement(Button, {
      type: "primary",
      onClick: () => onEdit(id)
    }, /*#__PURE__*/React.createElement(Icon, {
      name: "edit"
    }), " ", _t('Edit')) : null);
  };

  var DetailComponent = model.components && model.components.DataDetail || C('Model.DataDetail');
  return /*#__PURE__*/React.createElement(Page, {
    className: "xadmin-model-detail-".concat(model.key),
    title: model.title,
    onBack: () => history.back(),
    subTitle: /*#__PURE__*/React.createElement(ModelBlock, {
      name: "model.detail.nav"
    }),
    actions: renderActions()
  }, /*#__PURE__*/React.createElement(ModelBlock, {
    name: "model.detail.before"
  }), /*#__PURE__*/React.createElement(DetailComponent, {
    id: id
  }), /*#__PURE__*/React.createElement(ModelBlock, {
    name: "model.detail.after"
  }));
};

export { ModelListPage, ModelFormPage, ModelDetailPage };
