// xadmin-antd@3.3.2/es/form/components/base.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import _ from 'lodash';
import { Form, Input, Col } from 'antd';

var FieldGroup = (_ref) => {
  var {
    label,
    meta,
    input,
    field,
    tailLayout,
    children
  } = _ref;
  var attrs = field.attrs || {};
  var error = meta.touched && (meta.error || meta.submitError);
  var extra = field.description || field.help;
  var size = field.option && field.option.groupSize || attrs.groupSize || {
    labelCol: {
      xs: {
        span: 24
      },
      sm: {
        span: 5
      }
    },
    wrapperCol: {
      xs: {
        span: 24
      },
      sm: {
        span: 19,
        offset: tailLayout ? 5 : 0
      }
    }
  };

  var groupProps = _objectSpread(_objectSpread({
    extra
  }, size), {}, {
    required: field.required
  });

  if (error) {
    groupProps['validateStatus'] = 'error';

    if (_.isString(error)) {
      groupProps['help'] = error;
    }
  }

  var controlComponent = children ? children : /*#__PURE__*/React.createElement(Input, _extends({}, input, attrs));
  return /*#__PURE__*/React.createElement(Form.Item, _extends({
    label: label
  }, groupProps), controlComponent, field.formText ? /*#__PURE__*/React.createElement("span", {
    className: "ant-form-text"
  }, " ", field.formText) : null);
};

var FieldTableGroup = (_ref2) => {
  var {
    label,
    meta,
    input,
    field,
    nav,
    children
  } = _ref2;
  var attrs = field.attrs || {};
  var error = meta.touched && (meta.error || meta.submitError);
  var extra = field.description || field.help;
  var size = field.option && field.option.groupSize || attrs.groupSize || {
    labelCol: {
      xs: {
        span: 24
      },
      sm: {
        span: 5
      }
    },
    wrapperCol: {
      xs: {
        span: 24
      },
      sm: {
        span: 18
      }
    }
  };

  var groupProps = _objectSpread(_objectSpread({
    extra
  }, size), {}, {
    required: field.required
  });

  if (error) {
    groupProps['validateStatus'] = 'error';
    groupProps['help'] = error;
  }

  var controlComponent = children ? children : /*#__PURE__*/React.createElement(Input, _extends({}, input, attrs));
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Form.Item, _extends({
    label: label
  }, groupProps), nav, field.formText ? /*#__PURE__*/React.createElement("span", {
    className: "ant-form-text"
  }, " ", field.formText) : null), controlComponent);
};

var InlineGroup = (_ref3) => {
  var {
    label,
    meta,
    input,
    field,
    children
  } = _ref3;
  var attrs = field.attrs || {};
  var error = meta.touched && (meta.error || meta.submitError);
  var extra = field.description || field.help;
  var groupProps = {
    extra,
    required: field.required
  };

  if (error) {
    groupProps['validateStatus'] = 'error';
    groupProps['help'] = error;
  }

  var controlComponent = children ? children : /*#__PURE__*/React.createElement(Input, _extends({}, input, attrs));
  return /*#__PURE__*/React.createElement(Form.Item, groupProps, /*#__PURE__*/React.cloneElement(controlComponent, {
    inline: true
  }), field.formText ? /*#__PURE__*/React.createElement("span", {
    className: "ant-form-text"
  }, " ", field.formText) : null);
};

var SimpleGroup = (_ref4) => {
  var {
    label,
    meta,
    input,
    field,
    children
  } = _ref4;
  var attrs = field.attrs || {};
  var error = meta.touched && (meta.error || meta.submitError);
  var extra = field.description || field.help;
  var groupProps = {
    extra,
    required: field.required
  };

  if (error) {
    groupProps['validateStatus'] = 'error';
    groupProps['help'] = error;
  }

  var controlComponent = children ? children : /*#__PURE__*/React.createElement(Input, _extends({}, input, attrs));
  return /*#__PURE__*/React.createElement(Form.Item, _extends({
    label: label
  }, groupProps), controlComponent, field.formText ? /*#__PURE__*/React.createElement("span", {
    className: "ant-form-text"
  }, " ", field.formText) : null);
};

var ColGroup = (_ref5) => {
  var {
    label,
    meta,
    input,
    field,
    children
  } = _ref5;
  var attrs = field.attrs || {};
  var error = meta.touched && (meta.error || meta.submitError);
  var extra = field.description || field.help;
  var size = field.option && field.option.groupSize || attrs.groupSize || {
    labelCol: {
      xs: {
        span: 24
      },
      sm: {
        span: 8
      }
    },
    wrapperCol: {
      xs: {
        span: 24
      },
      sm: {
        span: 16
      }
    }
  };

  var groupProps = _objectSpread(_objectSpread({
    extra
  }, size), {}, {
    required: field.required
  });

  if (error) {
    groupProps['validateStatus'] = 'error';
    groupProps['help'] = error;
  }

  var controlComponent = children ? children : /*#__PURE__*/React.createElement(Input, _extends({}, input, attrs));
  return /*#__PURE__*/React.createElement(Col, {
    span: 8,
    xs: 24,
    sm: 12,
    md: 12,
    lg: 8,
    xl: 8,
    xxl: 6
  }, /*#__PURE__*/React.createElement(Form.Item, _extends({
    label: label
  }, groupProps), controlComponent, field.formText ? /*#__PURE__*/React.createElement("span", {
    className: "ant-form-text"
  }, field.formText) : null));
};

export { FieldGroup, FieldTableGroup, InlineGroup, ColGroup, SimpleGroup };
