// xadmin-antd@3.3.2/es/form/components/layout.js
import React from 'react';
import { Form, Button, Card, Space, Modal } from 'antd';
import app from 'xadmin';

var FormLayout = props => {
  var {
    children,
    invalid,
    handleSubmit,
    submitting
  } = props;
  var {
    _t
  } = app.context;
  var groupProps = {
    wrapperCol: {
      xs: {
        span: 24,
        offset: 0
      },
      sm: {
        span: 18,
        offset: 5
      }
    }
  };
  return /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement("form", {
    onSubmit: handleSubmit
  }, children, /*#__PURE__*/React.createElement(Form.Item, groupProps, /*#__PURE__*/React.createElement(Space, null, /*#__PURE__*/React.createElement(Button, {
    type: "primary",
    htmlType: "submit",
    loading: submitting,
    disabled: invalid
  }, _t('Save')), /*#__PURE__*/React.createElement(Button, {
    onClick: () => history.back()
  }, _t('Cancel'))))));
};

var ModalLayout = (_ref) => {
  var {
    children,
    invalid,
    handleSubmit,
    submitting,
    title,
    show,
    onClose,
    saveText
  } = _ref;
  var {
    _t
  } = app.context;
  return /*#__PURE__*/React.createElement(Modal, {
    visible: show,
    onClose: onClose,
    style: {
      maxWidth: 700
    },
    width: "95%",
    title: title,
    okText: saveText || _t('Save'),
    onOk: handleSubmit,
    okButtonProps: {
      disabled: invalid,
      loading: submitting
    },
    cancelText: _t('Cancel'),
    onCancel: onClose
  }, /*#__PURE__*/React.createElement(Form, {
    onSubmit: handleSubmit
  }, children));
};

export { FormLayout, ModalLayout };
