// xadmin@2.3.1/lib/rrs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _reactRedux = require("react-redux");

var _lodash = _interopRequireDefault(require("lodash"));

var _reactRouter = require("react-router");

var _redux = require("redux");

var _reduxSaga = _interopRequireDefault(require("redux-saga"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// redux app
var redux_app = {
  items: {
    reducers: {
      type: 'mapArray'
    },
    'redux/subscribe': {
      type: 'array'
    },
    'redux/on_create': {
      type: 'array'
    },
    'redux/middlewares': {
      type: 'array'
    },
    'redux/store_enhancers': {
      type: 'array'
    },
    'redux/reducer_enhance': {
      type: 'mapArray'
    }
  },
  context: app => (context, cb) => {
    var enhancers = [(0, _redux.applyMiddleware)(...app.get('redux/middlewares')), ...app.$('redux/store_enhancers')];

    var enhance_reducer = (key, reducer) => {
      var reducer_enhance = app.get('redux/reducer_enhance');

      if (reducer_enhance[key] !== undefined) {
        var reducers = [reducer, ...reducer_enhance[key]];
        return (state, action) => {
          return reducers.reduce((prev, reducer) => reducer(prev, action), state);
        };
      }

      return reducer;
    };

    var combine_reducer = (key, reducers) => {
      if (reducers.length > 1) {
        return (state, action) => {
          return reducers.reduce((prev, reducer) => reducer(prev, action), state);
        };
      } else {
        return reducers[0];
      }
    };

    var create_reducers = () => {
      var reducers_map = app.get('reducers');
      var reducers = {};

      for (var key in reducers_map) {
        reducers[key] = combine_reducer(key, reducers_map[key]);
      }

      return (0, _redux.combineReducers)(reducers);
    };

    var composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || _redux.compose; // create store

    var store = (0, _redux.createStore)(create_reducers(), context['initial_state'] || {}, composeEnhancers(...enhancers));
    cb(null, _objectSpread(_objectSpread({}, context), {}, {
      store
    }));
  },
  start: app => () => {
    // store change
    var {
      store
    } = app.context;
    var listeners = app.get('redux/subscribe');

    for (var listener of listeners) {
      store.subscribe(listener);
    }

    app.get('redux/on_create').forEach(callback => callback(store));
  }
}; // saga app

var sagaMiddleware = (0, _reduxSaga.default)();
var sage_app = {
  items: {
    effects: {
      type: 'array'
    }
  },
  context: app => (context, cb) => {
    // extend store
    var {
      store
    } = context;
    store.runSaga = sagaMiddleware.run; // start saga

    var effects = app.get('effects');
    effects.forEach(sagaMiddleware.run);
    cb(null, context);
  },
  'redux/middlewares': app => {
    return sagaMiddleware;
  }
}; // react & react-router app

var react_app = {
  items: {
    routers: {
      type: 'mapArray'
    },
    root_component: {
      type: 'array'
    }
  },
  context: app => (context, cb) => {
    app.go = uri => {
      app.context.router.push(uri);
    };

    var routerType = app.config('router') || 'browser';
    var router = typeof routerType === 'string' ? {
      browser: _reactRouter.browserHistory,
      hash: _reactRouter.hashHistory
    }[routerType] : routerType;
    cb(null, _objectSpread(_objectSpread({}, context), {}, {
      router
    }));
  },
  start: app => () => {
    // init container
    var {
      container = '#app'
    } = app.context;

    if (typeof container === 'string') {
      container = document.querySelector(container);
    }

    var rs = app.get('routers');

    var find_childs = path => {
      return (rs[path] || []).map(r => {
        var childs = find_childs((path == '@' ? '' : path) + r.path);
        return childs.length > 0 ? _objectSpread(_objectSpread({}, r), {}, {
          childRoutes: [...(r.childRoutes || []), ...childs]
        }) : r;
      });
    };

    var routers = find_childs('@');
    var root = app.get('root_component').reduce((children, render) => {
      return render(children);
    }, routers && routers.length ? /*#__PURE__*/_react.default.createElement(_reactRouter.Router, {
      history: app.context.router,
      routes: routers[0]
    }) : /*#__PURE__*/_react.default.createElement("span", null, "Please config routers or Main component."));

    _reactDom.default.render(root, container);
  }
}; // react-redux app

var react_redux_app = {
  root_component: app => children => /*#__PURE__*/_react.default.createElement(_reactRedux.Provider, {
    store: app.context.store
  }, children),
  hooks: app => ({
    'redux': (props, select) => {
      var store = app.context.store;
      var {
        getState,
        dispatch,
        subscribe
      } = store;

      if (select) {
        var state = getState();

        var [values, setValues] = _react.default.useState(select(state) || {});

        var lastValues = _react.default.useRef();

        lastValues.current = values;

        var updateState = () => {
          var newValues = select(getState());

          if (!_lodash.default.isEqual(lastValues.current, newValues)) {
            setValues(newValues);
          }
        };

        _react.default.useEffect(() => {
          return subscribe(updateState);
        }, []);

        return _objectSpread(_objectSpread({}, props), {}, {
          store,
          dispatch: dispatch,
          state
        }, values);
      } else {
        return _objectSpread(_objectSpread({}, props), {}, {
          store,
          dispatch: dispatch,
          state: getState()
        });
      }
    }
  })
};

var _default = app => {
  return app.use(redux_app).use(sage_app).use(react_app).use(react_redux_app);
};

exports.default = _default;
