// xadmin-antd@3.3.2/lib/model/components/DetailModal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _lodash = _interopRequireDefault(require("lodash"));

var _antd = require("antd");

var _xadminUi = require("xadmin-ui");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class DetailModal extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      show: false
    });
  }

  showModal() {
    this.setState({
      show: true
    });
  }

  hideModal() {
    this.setState({
      show: false
    });
  }

  render() {
    var {
      id,
      children
    } = this.props;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("a", {
      style: {
        cursor: 'pointer'
      },
      onClick: this.showModal
    }, children), this.state.show && /*#__PURE__*/_react.default.createElement(_antd.Modal, {
      visible: this.state.show,
      onCancel: this.hideModal,
      title: children
    }, /*#__PURE__*/_react.default.createElement(_xadminUi.C, {
      is: "Model.DataDetail",
      id: id
    })));
  }

}

var _default = DetailModal;
exports.default = _default;
