// xadmin-model@3.3.3/es/search.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React from 'react';
import _ from 'lodash';
import { C } from 'xadmin-ui';
import { use } from 'xadmin';
export default {
  name: 'xadmin.model.search',
  blocks: {
    'model.list.nav': props => /*#__PURE__*/React.createElement(C, _extends({
      is: "Model.SearchBar",
      key: "searchBar"
    }, props))
  },
  hooks: {
    'model.searchbar': () => {
      var [searchValue, setSearch] = React.useState(null);
      var setSearchFilter = use('model.setter', 'where', 'searchbar');
      var {
        model
      } = use('model');
      var searchFields = model.searchFields;
      var searchTitles = model.searchFields && model.searchFields.map(field => model.properties[field].title || field);
      var onSearch = React.useCallback(search => {
        if (search) {
          var searchs = model.searchFields.map(field => {
            return {
              [field]: {
                like: search
              }
            };
          });

          if (searchs.length > 1) {
            setSearchFilter({
              or: searchs
            });
          } else if (searchs.length > 0) {
            setSearchFilter(searchs[0]);
          }
        } else {
          setSearchFilter(null);
        }

        setSearch(search);
      }, [model.searchFields, setSearch, setSearchFilter]);
      return {
        searchValue,
        searchFields,
        searchTitles,
        onSearch
      };
    }
  }
};
