// xadmin-antd@3.3.2/es/model/actions/BatchChange.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import _ from 'lodash';
import { Modal, Menu, Form } from 'antd';
import { SchemaForm } from 'xadmin-form';
import { app, use } from 'xadmin';

var BatchChangeBtn = props => {
  var {
    _t
  } = app.context;
  var [show, setShow] = React.useState(false);
  var {
    canEdit,
    fields,
    loading,
    onBatchChange
  } = use('actons.batch_change');
  var {
    selected
  } = use('model.select');
  var {
    model
  } = use('model');

  var onClose = () => setShow(false);

  var renderModel = () => {
    var fs = fields.map(f => f.split('.')[0]);
    return /*#__PURE__*/React.createElement(SchemaForm, {
      key: "actions_batch_change_form",
      formKey: "model_batch.".concat(model.key),
      schema: _.omit(_objectSpread(_objectSpread({}, model), {}, {
        properties: _.pick(model.properties, fs),
        form: model.form !== undefined ? fs.map(name => _.find(model.form, f => f && f.key == name) || name) : ['*']
      }), 'required'),
      onSubmit: onBatchChange,
      onSubmitSuccess: onClose,
      onClose: onClose
    }, (_ref) => {
      var {
        children,
        invalid,
        handleSubmit,
        submitting,
        onClose
      } = _ref;
      return /*#__PURE__*/React.createElement(Modal, {
        key: "actions_batch_change_modal",
        visible: show,
        onClose: onClose,
        title: _t('Please input the value to batch change items'),
        okText: _t('Change'),
        onOk: handleSubmit,
        okButtonDisabled: invalid || submitting,
        cancelText: _t('Cancel'),
        onCancel: onClose,
        okButtonProps: {
          loading
        }
      }, /*#__PURE__*/React.createElement(Form, {
        onSubmit: handleSubmit
      }, children));
    });
  };

  return canEdit && fields.length > 0 ? [/*#__PURE__*/React.createElement(Menu.Item, _extends({}, props, {
    key: "actions_batch_change",
    onClick: e => {
      props.onClick && props.onClick(e);
      setShow(true);
    },
    disabled: selected.length == 0
  }), _t('Batch Change Items')), selected.length > 0 ? renderModel() : null] : null;
};

export default BatchChangeBtn;
