// xadmin-antd@3.3.2/lib/form/components/Checkbox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var CheckboxComponent = (_ref) => {
  var {
    input,
    label,
    field,
    meta,
    group: FieldGroup
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(FieldGroup, {
    meta: meta,
    input: input,
    field: field,
    tailLayout: true
  }, /*#__PURE__*/_react.default.createElement(_antd.Checkbox, _extends({
    checked: !!input.value
  }, input, field.attrs), label));
};

CheckboxComponent.useGroup = false;
var _default = CheckboxComponent;
exports.default = _default;
