// xadmin-antd@3.3.2/lib/form/components/DatePicker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var _moment = _interopRequireDefault(require("moment"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class DatePickerInput extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "onChange", value => {
      var {
        input
      } = this.props;
      input.onChange(this.getValue(value));
    });
  }

  getValue(value) {
    var {
      field
    } = this.props;
    var format = field.datetimeFormat || 'YYYY-MM-DD';
    return value && value.format(format);
  }

  render() {
    var {
      input,
      field
    } = this.props;
    var format = field.datetimeFormat || 'YYYY-MM-DD';
    return /*#__PURE__*/_react.default.createElement(_antd.DatePicker, _extends({
      allowClear: true
    }, input, {
      onChange: this.onChange,
      value: input.value ? (0, _moment.default)(input.value) : null,
      format: format
    }));
  }

}

exports.default = DatePickerInput;
