// xadmin-antd@3.3.2/lib/form/components/layout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalLayout = exports.FormLayout = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var _xadmin = _interopRequireDefault(require("xadmin"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FormLayout = props => {
  var {
    children,
    invalid,
    handleSubmit,
    submitting
  } = props;
  var {
    _t
  } = _xadmin.default.context;
  var groupProps = {
    wrapperCol: {
      xs: {
        span: 24,
        offset: 0
      },
      sm: {
        span: 18,
        offset: 5
      }
    }
  };
  return /*#__PURE__*/_react.default.createElement(_antd.Card, null, /*#__PURE__*/_react.default.createElement("form", {
    onSubmit: handleSubmit
  }, children, /*#__PURE__*/_react.default.createElement(_antd.Form.Item, groupProps, /*#__PURE__*/_react.default.createElement(_antd.Space, null, /*#__PURE__*/_react.default.createElement(_antd.Button, {
    type: "primary",
    htmlType: "submit",
    loading: submitting,
    disabled: invalid
  }, _t('Save')), /*#__PURE__*/_react.default.createElement(_antd.Button, {
    onClick: () => history.back()
  }, _t('Cancel'))))));
};

exports.FormLayout = FormLayout;

var ModalLayout = (_ref) => {
  var {
    children,
    invalid,
    handleSubmit,
    submitting,
    title,
    show,
    onClose,
    saveText
  } = _ref;
  var {
    _t
  } = _xadmin.default.context;
  return /*#__PURE__*/_react.default.createElement(_antd.Modal, {
    visible: show,
    onClose: onClose,
    style: {
      maxWidth: 700
    },
    width: "95%",
    title: title,
    okText: saveText || _t('Save'),
    onOk: handleSubmit,
    okButtonProps: {
      disabled: invalid,
      loading: submitting
    },
    cancelText: _t('Cancel'),
    onCancel: onClose
  }, /*#__PURE__*/_react.default.createElement(_antd.Form, {
    onSubmit: handleSubmit
  }, children));
};

exports.ModalLayout = ModalLayout;
