// xadmin-auth@3.3.2/lib/effects.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _effects = require("redux-saga/effects");

var _xadmin = require("xadmin");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function* handle_get_userinfo(_ref) {
  var {
    type
  } = _ref;

  try {
    var user = yield (0, _xadmin.api)({
      name: 'auth'
    }).get('user');
    yield (0, _effects.put)({
      type,
      payload: user,
      success: true
    });
  } catch (err) {
    _xadmin.app.error(err);
  }
}

function* handle_user_signin(_ref2) {
  var {
    payload: user
  } = _ref2;

  if (user.name == undefined && user.username == undefined) {
    yield (0, _effects.put)({
      type: '@@xadmin/GET_USER_INFO'
    });
  }
}

function* handle_user_signout() {
  var {
    _t
  } = _xadmin.app.context;

  try {
    yield (0, _xadmin.api)({
      name: 'auth/logout'
    }).save({});
    yield (0, _effects.put)({
      type: '@@xadmin/AUTH_SIGN_OUT_FINISH'
    });
    yield (0, _effects.put)({
      type: '@@xadmin/ADD_NOTICE',
      payload: {
        type: 'success',
        headline: 'Success',
        message: _t('Successfully logged out')
      }
    });
  } catch (err) {
    _xadmin.app.error(err);
  }
}

function* handle_verify_email(_ref3) {
  var {
    payload
  } = _ref3;
  var {
    _t
  } = _xadmin.app.context;

  try {
    yield (0, _xadmin.api)({
      name: 'auth/registration/verify-email'
    }).save(_objectSpread({
      language: 'zh_CN'
    }, payload));
    yield (0, _effects.put)({
      type: '@@xadmin/ADD_NOTICE',
      payload: {
        type: 'success',
        headline: 'Success',
        message: _t('Send verify code to your email, please check')
      }
    });
  } catch (err) {
    _xadmin.app.error(err);
  }
}

function* _default() {
  yield (0, _effects.all)([(0, _effects.takeEvery)(action => action.type == '@@xadmin/GET_USER_INFO' && action.success != true, handle_get_userinfo), (0, _effects.takeEvery)(action => action.type == '@@xadmin/AUTH_VERIFY_EMAIL', handle_verify_email), (0, _effects.takeEvery)(action => action.type == '@@xadmin/AUTH_SIGN_IN', handle_user_signin), (0, _effects.takeEvery)(action => action.type == '@@xadmin/AUTH_SIGN_OUT', handle_user_signout)]);
}
