// xadmin-auth@3.3.2/lib/hooks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _react = _interopRequireDefault(require("react"));

var _localforage = _interopRequireDefault(require("localforage"));

var _jsCookie = _interopRequireDefault(require("js-cookie"));

var _xadmin = require("xadmin");

var _xadminI18n = require("xadmin-i18n");

var _context = require("./context");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var _default = {
  'auth.user': () => _react.default.useContext(_context.UserContext),
  'auth.context': () => {
    var {
      auth
    } = _xadmin.app.load_dict('config');

    var [init, setInit] = _react.default.useState(false);

    var [user, setUserInfo] = _react.default.useState(null);

    var setUser = user => {
      _xadmin.app.context.user = user;
      setUserInfo(user);
    };

    _react.default.useEffect(() => {
      var loadUser = /*#__PURE__*/function () {
        var _ref = _asyncToGenerator(function* () {
          var user = null;

          if (auth.persist_type == 'localforage') {
            try {
              var value = yield _localforage.default.getItem('user');

              if (value) {
                user = JSON.parse(value);
              }
            } catch (error) {
              yield _localforage.default.removeItem('user');
            }
          } else if (auth.persist_type == 'session-storage') {
            var _value = sessionStorage.getItem('user');

            if (_value) {
              user = JSON.parse(_value);
            }
          } else if (auth.persist_type == 'cookie') {
            user = _jsCookie.default.getJSON('user');
          }

          if (!_lodash.default.isEmpty(user)) {
            setUser(user);

            try {
              var userInfo = yield (0, _xadmin.api)({
                name: 'auth'
              }).get('user');

              if (!_lodash.default.isEmpty(userInfo)) {
                setUser(_objectSpread(_objectSpread({}, user), userInfo));
              }
            } catch (err) {}
          }

          setTimeout(() => setInit(true), 0);
        });

        return function loadUser() {
          return _ref.apply(this, arguments);
        };
      }();

      loadUser();
    }, []);

    _react.default.useEffect(() => {
      var saveUser = /*#__PURE__*/function () {
        var _ref2 = _asyncToGenerator(function* () {
          if (auth.persist_type == 'localforage') {
            if (user) {
              yield _localforage.default.setItem('user', JSON.stringify(user));
            } else {
              yield _localforage.default.removeItem('user');
            }
          } else if (auth.persist_type == 'session-storage') {
            if (user) {
              sessionStorage.setItem('user', JSON.stringify(user));
            } else {
              sessionStorage.removeItem('user');
            }
          } else if (auth.persist_type == 'cookie') {
            if (user) {
              _jsCookie.default.set('user', user, auth.userinfo_timeout ? {
                expires: auth.userinfo_timeout
              } : {});
            } else {
              _jsCookie.default.remove('user');
            }
          }
        });

        return function saveUser() {
          return _ref2.apply(this, arguments);
        };
      }();

      saveUser();
    }, [user]);

    return {
      show: init,
      user,
      setUser
    };
  },
  'auth.login': () => {
    var location = (0, _xadmin.use)('location');
    var query = (0, _xadmin.use)('query');
    var navigate = (0, _xadmin.use)('navigate');
    var message = (0, _xadmin.use)('message');
    var {
      setUser
    } = (0, _xadmin.use)('auth.user');

    var onSignIn = /*#__PURE__*/function () {
      var _ref3 = _asyncToGenerator(function* (item) {
        try {
          var _location$state, _location$state$from;

          var user = yield (0, _xadmin.api)({
            name: 'auth/login'
          }).save(item);
          setUser(_lodash.default.omit(user, 'password'));

          if (user.name == undefined && user.username == undefined) {
            try {
              var userInfo = yield (0, _xadmin.api)({
                name: 'auth'
              }).get('user');
              setUser(_objectSpread(_objectSpread({}, user), userInfo));
            } catch (err) {}
          }

          navigate(((_location$state = location.state) === null || _location$state === void 0 ? void 0 : (_location$state$from = _location$state.from) === null || _location$state$from === void 0 ? void 0 : _location$state$from.pathname) || '/app/', {
            replace: true
          });
        } catch (error) {
          console.log(error);
          throw error.json && error.json.non_field_errors == undefined ? error.json : {
            password: (0, _xadminI18n._t)('Incorrect username or password')
          };
        }
      });

      return function onSignIn(_x) {
        return _ref3.apply(this, arguments);
      };
    }();

    _react.default.useEffect(() => {
      if (query.verifyEmail == 'success') {
        message && message.success((0, _xadminI18n._t)('Verify email success, please login'));
      }
    }, []);

    return {
      onSignIn
    };
  },
  'auth.logout': () => {
    var {
      setUser
    } = (0, _xadmin.use)('auth.user');
    var message = (0, _xadmin.use)('message');

    var onLogout = _react.default.useCallback( /*#__PURE__*/_asyncToGenerator(function* () {
      try {
        yield (0, _xadmin.api)({
          name: 'auth/logout'
        }).save({});
        message && message.success((0, _xadminI18n._t)('Successfully logged out'));
        setUser(null);
      } catch (err) {
        _xadmin.app.error(err);
      }
    }), []);

    return {
      onLogout
    };
  },
  'auth.sign_up': () => {
    var navigate = (0, _xadmin.use)('navigate');
    var message = (0, _xadmin.use)('message');

    var onSuccess = resp => {
      if (resp.key) {
        (0, _xadmin.api)({
          name: 'auth/registration/verify-email'
        }).save(_objectSpread({
          language: 'zh_CN'
        }, resp)).then(() => {
          message && message.success((0, _xadminI18n._t)('Send verify code to your email, please check'));
        });
      } else if (resp.detail) {
        message && message.success(resp.detail);
      }

      navigate('/login');
    };

    return {
      onSuccess
    };
  },
  'auth.forget_password': () => {
    var navigate = (0, _xadmin.use)('navigate');

    var onSuccess = user => {
      navigate('/login');
    };

    return {
      onSuccess
    };
  },
  'auth.change_password': () => {
    var navigate = (0, _xadmin.use)('navigate');
    var message = (0, _xadmin.use)('message');

    var onChange = /*#__PURE__*/function () {
      var _ref5 = _asyncToGenerator(function* (item) {
        try {
          yield (0, _xadmin.api)({
            name: 'user/password'
          }).save(item);
          message && message.success((0, _xadminI18n._t)('Change password success'));
          navigate('/app/');
        } catch (error) {
          throw new Error({
            old_password: (0, _xadminI18n._t)('Incorrect old password')
          });
        }
      });

      return function onChange(_x2) {
        return _ref5.apply(this, arguments);
      };
    }();

    return {
      onChange
    };
  },
  'auth.reset_password': () => {
    var navigate = (0, _xadmin.use)('navigate');

    var onSuccess = user => {
      navigate('/login');
    };

    return {
      onSuccess
    };
  }
};
exports.default = _default;
