// xadmin-model@3.3.3/lib/filter/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BaseFilter = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _lodash = _interopRequireDefault(require("lodash"));

var _xadmin = _interopRequireWildcard(require("xadmin"));

var _xadminForm = require("xadmin-form");

var _xadminUi = require("xadmin-ui");

var _utils = require("../utils");

var _filters = _interopRequireDefault(require("./filters"));

var _recoil = require("recoil");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var convert = (schema, options) => {
  var opts = options || {};
  return _xadmin.default.load_list('filter_converter').reduce((prve, converter) => {
    return converter(prve, schema, opts);
  }, {});
};

var FilterForm = props => {
  var {
    component: FilterComponent
  } = props,
      filterForm = _objectWithoutProperties(props, ["component"]);

  var {
    resetFilter
  } = (0, _xadmin.use)('model.list.filter');
  return /*#__PURE__*/_react.default.createElement(FilterComponent, _extends({}, filterForm, {
    resetFilter: resetFilter
  }));
};

var BaseFilter = props => {
  var {
    name,
    component: _component,
    group,
    fieldProps
  } = props;
  var {
    model,
    atoms
  } = (0, _xadmin.use)('model');
  var [data, setFilters] = (0, _recoil.useRecoilState)(atoms.where('filters'));

  var {
    filters,
    options,
    formKey
  } = _react.default.useMemo(() => {
    var formKey = "filter.".concat(model.key || model.name);
    var filter = model.filters && model.filters[name];
    var fields, options;

    if (_lodash.default.isArray(filter)) {
      options = {};
      fields = filter;
    } else if (_lodash.default.isPlainObject(filter) && _lodash.default.isArray(filter.fields)) {
      fields = filter.fields;
      options = _lodash.default.omit(filter, 'fields');
    } else {
      return {
        filters: [],
        formKey,
        options: {}
      };
    }

    var filters = fields.map(field => {
      var key = typeof field == 'string' ? field : field.key;
      var schema = (0, _utils.getFieldProp)(model, key);
      return schema ? {
        key,
        schema,
        field: typeof field == 'string' ? {} : field
      } : null;
    }).filter(Boolean);
    return {
      filters,
      formKey,
      options
    };
  }, [model, name]);

  var fields = _react.default.useMemo(() => filters.map(filter => {
    var field = convert(filter.schema, {
      key: filter.key
    });
    return _lodash.default.merge(field, filter.field, fieldProps);
  }), [filters, fieldProps]);

  var onSubmit = _react.default.useCallback(values => {
    var where = Object.keys(values).reduce((prev, key) => {
      if (!_lodash.default.isNil(values[key])) {
        prev[key] = values[key];
      } else {
        prev = _lodash.default.omit(prev, key);
      }

      return prev;
    }, _objectSpread({}, data));
    setFilters(where);
  }, [model, data, setFilters]);

  var onChange = options && options.submitOnChange == true ? onSubmit : undefined;
  return fields && filters.length ? /*#__PURE__*/_react.default.createElement(_xadminForm.Form, {
    onSubmit: onSubmit,
    onChange: onChange,
    fields: fields,
    component: props => /*#__PURE__*/_react.default.createElement(FilterForm, _extends({}, props, {
      component: _component
    })),
    group: group,
    initialValues: data || {},
    options: options
  }) : null;
};

exports.BaseFilter = BaseFilter;

var filter = (name, componentName, groupName) => (_ref) => {
  var {
    model
  } = _ref;
  return model && model.filters && model.filters[name] ? /*#__PURE__*/_react.default.createElement(BaseFilter, {
    name: name,
    component: (0, _xadminUi.C)(componentName),
    group: groupName && (0, _xadminUi.C)(groupName)
  }) : null;
};

var _default = {
  name: 'xadmin.filter',
  components: {
    'Model.BaseFilter': BaseFilter
  },
  blocks: {
    'model.list.nav': filter('navform', 'Filter.NavForm', 'Form.InlineGroup'),
    'model.list.navbtn': filter('nav', 'Filter.Modal'),
    'model.list.submenu': filter('submenu', 'Filter.Submenu', 'Form.ColGroup'),
    'model.list.sidemenu': filter('sidemenu', 'Filter.Form', 'Form.SimpleGroup')
  },
  hooks: {
    'model.list.filter': () => {
      var {
        model,
        atoms
      } = (0, _xadmin.use)('model');
      var {
        form
      } = (0, _xadmin.use)('form');
      var setFilters = (0, _recoil.useSetRecoilState)(atoms.where('filters'));

      var resetFilter = _react.default.useCallback(() => {
        var initial = _lodash.default.isFunction(model.initialValues) ? model.initialValues() : model.initialValues;
        var where = initial && initial.wheres && initial.wheres.filters || {};
        form.reset(where);
        setFilters(where);
      }, [model, setFilters]);

      _react.default.useEffect(() => {
        if (model.filterDefault) {
          var values = _lodash.default.isFunction(model.filterDefault) ? model.filterDefault() : model.filterDefault;
          form.reset(values);
        }
      }, [model.filterDefault]);

      return {
        resetFilter
      };
    }
  },
  filter_converter: _filters.default
};
exports.default = _default;
