// @airiot/devtool@1.0.7/config/webpackDevServer.config.js
const paths = require('./paths');
const path = require('path');
const fs = require('fs');

const envs = process.env

module.exports = host => {
  
  const proxy = {
    '/rest': {
      target: host,
      changeOrigin: true
    },
    '/ws': {
      ws: true,
      target: host,
      changeOrigin: true
    }
  }
  
  if(envs.CLOUD_URL) {
    proxy['/rest/cloud'] = {
      target: envs.CLOUD_URL,
      pathRewrite: { '^/rest/cloud': '' },
      changeOrigin: true
    }
  }
  
  proxy['/node_modules'] = {
    target: host,
    changeOrigin: true
  }
  
  return {
    historyApiFallback: {
      disableDotRule: true,
    },
    inline: true,
    compress: true,
    clientLogLevel: 'none',
    watchContentBase: true,
    hot: true,
    contentBase: [
      paths.buildPath,
      paths.devtoolBuild,
      paths.appNodeModules
    ].filter(Boolean),
    before(app, server) {
      if (fs.existsSync(paths.proxySetup)) {
        require(paths.proxySetup)(app);
      }
    },
    proxy
  }
}
